core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -1118,33 +1171,38 @@ class GroupCoordinator(val brokerId: Int,
     group.removeStaticMember(member.groupInstanceId)
 
     group.currentState match {
-      case Dead | Empty =>
-      case Stable | CompletingRebalance => maybePrepareRebalance(group, reason)
-      case PreparingRebalance => joinPurgatory.checkAndComplete(GroupKey(group.groupId))
+      case Dead | Empty => None
+      case Stable | CompletingRebalance =>
+        maybePrepareRebalance(group, reason)
+        None
+      case PreparingRebalance => Some(GroupKey(group.groupId))
     }
   }
 
-  private def removePendingMemberAndUpdateGroup(group: GroupMetadata, memberId: String): Unit = {
+  /**
+   * remove the pending member and then return the group key which is in PreparingRebalance,
+   * @param group group
+   * @param memberId member id
+   * @return group key if it is in PreparingRebalance. Otherwise, None
+   */
+  private def removePendingMemberAndUpdateGroup(group: GroupMetadata, memberId: String): Option[GroupKey] = {
     group.removePendingMember(memberId)
 
-    if (group.is(PreparingRebalance)) {
-      joinPurgatory.checkAndComplete(GroupKey(group.groupId))
-    }
-  }
-
-  def tryCompleteJoin(group: GroupMetadata, forceComplete: () => Boolean) = {
-    group.inLock {
-      if (group.hasAllMembersJoined)
-        forceComplete()
-      else false
-    }
+    if (group.is(PreparingRebalance)) Some(GroupKey(group.groupId))
+    else None
   }
 
   def onExpireJoin(): Unit = {
     // TODO: add metrics for restabilize timeouts
   }
 
-  def onCompleteJoin(group: GroupMetadata): Unit = {
+  /**
+   * @return Returning a map of successfully appended topic partitions and a flag indicting whether the HWM has been
+   *         incremented. If the caller no longer passed in completeDelayedRequests, the caller is expected to complete
```

Comment:
" If the caller no longer passed in completeDelayedRequests" => There is no completeDelayedRequests passed in.